### PR TITLE
aruha-526: Include Timelines into cache

### PR DESCRIPTION
**Pull request by adyach**

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java [hunk 1]
```diff
@@ -20,75 +13,111 @@
...
+        this.timelineSync = timelineSync;
+        this.timelineRegistrations = new ConcurrentHashMap<>();
         if (null != cacheSync) {
             this.cacheSync.getListenable().addListener((curator, event) -> this.onZkEvent(event));
         }
-        preloadEventTypes(eventTypeRepository);
+        preloadEventTypes(eventTypeRepository, timelineRepository);
+    }
+
+    @PreDestroy
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java]: I thought `EventTypeCache` is never destroyed?

Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java]: @v-stepanov it is called when application context is going to be destroyed.
http://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-spring-application.html#boot-features-application-exit

Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java]: But is this a real case for us? Why do we need it?

File: src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java [hunk 2]
```diff
@@ -31,6 +37,10 @@ public TimelineDbRepository(final JdbcTemplate jdbcTemplate, final ObjectMapper
         super(jdbcTemplate, jsonMapper);
     }
 
+    public List<Timeline> list() {
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: I think it's very inconsistent that we have methods named `list` and `listTimelines` here.

File: src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java (hunk 1 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java]: If you acquire resource you have to clean it up when you will finish using it. i do not think there is much sense in doing it here. I will remove it if it is mess for you.

File: src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java (hunk 2 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: What is your proposal? Name both listTimelines?

Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: Yes, i think it's a good solution.

Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: b385578

File: src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java (hunk 1 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java]: b385578

File: src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java [hunk 3]
```diff
@@ -146,23 +172,19 @@ public EventTypeValidator getValidator(final String name) throws InternalNakadiE
...
-        }
-
-        final PathChildrenCache cacheSync = new PathChildrenCache(zkClient, ZKNODE_PATH, false);
-
-        // It is important to preload all data before specifying callback for updates, because otherwise preload won't
-        // give any effect - all changes will be removed.
-        cacheSync.start(PathChildrenCache.StartMode.BUILD_INITIAL_CACHE);
+    public Optional<Timeline> getActiveTimeline(final String name) throws InternalNakadiException,
+            NoSuchEventTypeException {
+        return getCached(name).map(CachedValue::getTimelines)
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java]: Wouldn't it be better to store the active timeline directly so not to map+filter every time this method is called?

**Comment by rcillo** [comment]
@adyach as we discussed offline, all cases are covered here. It seems good 👍

**Comment by rcillo** [comment]
:shipit:

File: src/acceptance-test/java/org/zalando/nakadi/repository/db/EventTypeCacheTestAT.java [hunk 4]
```diff
@@ -129,13 +145,106 @@ public void invalidateCacheOnUpdate() throws Exception {
...
+
+    private List<Timeline> getMockedTimelines(final String etName) {
+        final Timeline t1 = new Timeline(etName, 0, new Storage(), "topic", new Date());
+        final Timeline t2 = new Timeline(etName, 1, new Storage(), "topic", new Date(System.currentTimeMillis() + 200));
+        t2.setSwitchedAt(new Date(System.currentTimeMillis() + 300));
+        final Timeline t3 = new Timeline(etName, 2, new Storage(), "topic", new Date(System.currentTimeMillis() + 500));
+        return ImmutableList.of(t1, t2, t3);
+    }
+
+    private class TestTimelineSync implements TimelineSync {
```
Comment (antban) [inline src/acceptance-test/java/org/zalando/nakadi/repository/db/EventTypeCacheTestAT.java]: Isn't it simpler to create mock?

Comment (adyach) [inline src/acceptance-test/java/org/zalando/nakadi/repository/db/EventTypeCacheTestAT.java]: This stub is used only for specific case when I want to call subscribes in order to invalidate cache.

**Comment by adyach** [comment]
:shipit:

**Pull request merged**